File: src/main/java/net/miarma/mkernel/commands/misc/WarpCommand.java
package net.miarma.mkernel.commands.misc;

import java.io.File;
import java.io.IOException;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

import org.bukkit.Location;
import org.bukkit.World;
import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.configuration.file.YamlConfiguration;

import dev.jorel.commandapi.CommandAPICommand;
import net.miarma.mkernel.MKernel;
import net.miarma.mkernel.common.minecraft.Warp;
import net.miarma.mkernel.config.CommandWrapper;
import net.miarma.mkernel.config.providers.CommandProvider;
import net.miarma.mkernel.config.providers.ConfigProvider;
import net.miarma.mkernel.config.providers.MessageProvider;
import net.miarma.mkernel.util.MessageUtil;

public class WarpCommand {
    public static void register() {
        CommandWrapper warpCmd = CommandProvider.getWarpCommand();
        CommandWrapper addSubCmd = warpCmd.getSubcommands()[0];
        CommandWrapper removeSubCmd = warpCmd.getSubcommands()[1];
        new CommandAPICommand(warpCmd.getName())
            .withPermission(warpCmd.getPermission().base())
            .withFullDescription(warpCmd.getDescription())
            .withShortDescription(warpCmd.getDescription())
            .withUsage(warpCmd.getUsage())
            .executesPlayer((sender, args) -> {
                File f = new File(MKernel.PLUGIN.getDataFolder().getAbsolutePath(), "warps/"
                        + sender.getName() + ".yml");
                FileConfiguration c = YamlConfiguration.loadConfiguration(f);

                Set<Warp> warps = c.getKeys(false).stream()
                        .map(alias -> Warp.fromFile(c, alias))
                        .collect(Collectors.toSet());

                if (warps.isEmpty()) {
                    MessageUtil.sendMessage(sender, warpCmd.getMessages()[0], true);
                } else {
                    String warpList = warps.stream()
                            .map(Warp::toFormattedMessage)
                            .collect(Collectors.joining("\n"));
                    warpList = MessageUtil.formatMessage(warpList, false);

                    MessageUtil.sendMessage(sender, warpCmd.getMessages()[1], true,
                                                List.of("%warps%"), List.of(warpList));
                }

            })
            .withSubcommand(
                new CommandAPICommand(addSubCmd.getName())
                    .withPermission(addSubCmd.getPermission().base())
                    .withFullDescription(addSubCmd.getDescription())
                    .withShortDescription(addSubCmd.getDescription())
                    .withArguments(CommandProvider.Arguments.warpName())
                    .withUsage(addSubCmd.getUsage())
                    .executesPlayer((sender, args) -> {
                        File f = new File(MKernel.PLUGIN.getDataFolder().getAbsolutePath(), "warps/"
                                + sender.getName() + ".yml");
                        FileConfiguration c = YamlConfiguration.loadConfiguration(f);

                        if (c.getKeys(false).size() >= ConfigProvider.Values.getMaxWarps()) {
                            MessageUtil.sendMessage(sender, MessageProvider.Errors.maxWarpsReached(), true);
                            return;
                        }

                        if(c.contains(args.getRaw(0))) {
                            MessageUtil.sendMessage(sender, addSubCmd.getMessages()[1], true,
                                                        List.of("%warp%"), List.of(args.getRaw(0)));
                            return;
                        }

                        String warpName = args.getRaw(0);
                        Location loc = sender.getLocation();
                        World world = sender.getWorld();
                        Warp warp = Warp.of(
                            warpName,
                            Math.round(loc.getX()),
                            Math.round(loc.getY()),
                            Math.round(loc.getZ()),
                            world.getName()
                        );
                        Warp.toFile(c, warp);

                        try {
                            c.save(f);
                        } catch (IOException e) {
                            throw new RuntimeException(e);
                        }

                        MessageUtil.sendMessage(sender, addSubCmd.getMessages()[0], true,
                                                    List.of("%warp%"), List.of(warpName));
                    })
            )
            .withSubcommand(
                new CommandAPICommand(removeSubCmd.getName())
                    .withPermission(removeSubCmd.getPermission().base())
                    .withFullDescription(removeSubCmd.getDescription())
                    .withShortDescription(removeSubCmd.getDescription())
                    .withArguments(CommandProvider.Arguments.warps())
                    .withUsage(removeSubCmd.getUsage())
                    .executes((sender, args) -> {
                        File f = new File(MKernel.PLUGIN.getDataFolder().getAbsolutePath(), "warps/"
                                + sender.getName() + ".yml");
                        FileConfiguration c = YamlConfiguration.loadConfiguration(f);

                        String warpName = args.getRaw(0);
                        if (c.contains(warpName)) {
                            c.set(warpName, null);
                            try {
                                c.save(f);
                            } catch (IOException e) {
                                MKernel.LOGGER.severe("Error al guardar el archivo de warps de " + sender.getName());
                            }
                            MessageUtil.sendMessage(sender, removeSubCmd.getMessages()[0], true,
                                                       List.of("%warp%"), List.of(warpName));
                        } else {
                            MessageUtil.sendMessage(sender, removeSubCmd.getMessages()[1], true,
                                                       List.of("%warp%"), List.of(warpName));
                        }
                    })
            )
            .register();

    }
}

File: src/main/kotlin/net/miarma/mkernel/command/misc/WarpCommand.kt
package net.miarma.mkernel.command.misc

import com.google.inject.Inject
import com.google.inject.Singleton
import dev.jorel.commandapi.CommandAPICommand
import dev.jorel.commandapi.arguments.StringArgument
import dev.jorel.commandapi.kotlindsl.commandAPICommand
import dev.jorel.commandapi.kotlindsl.playerExecutor
import net.miarma.mkernel.common.service.impl.ConfigService
import net.miarma.mkernel.common.service.impl.DatabaseService
import net.miarma.mkernel.common.service.impl.MessageService

@Singleton
class WarpCommand @Inject constructor(
    private val configService: ConfigService,
    private val messageService: MessageService,
    private val databaseService: DatabaseService
) {
    fun register() {
        commandAPICommand(configService.getString("commands.warp.name")) {
            withPermission(configService.getString("commands.warp.permission"))
            withFullDescription(configService.getString("commands.warp.description"))
            withUsage(configService.getString("commands.warp.usage"))
            playerExecutor { sender, _ ->
                databaseService.getWarpObjects(sender) { warps ->
                    if (warps.isEmpty()) {
                        messageService.builder(configService.getString("commands.warp.messages.noWarpsStored")).withPrefix().send(sender)
                    } else {
                        val warpList = warps.joinToString("\n") { it.toFormattedMessage() }
                        messageService.builder(configService.getString("commands.warp.messages.warpList")).withPrefix().tag("warps", warpList).send(sender)
                    }
                }
            }
            withSubcommand(CommandAPICommand(configService.getString("commands.warp.subcommands.add.name")).apply {
                withPermission(configService.getString("commands.warp.subcommands.add.permission"))
                withArguments(StringArgument(configService.getString("arguments.warpName")))
                withUsage(configService.getString("commands.warp.subcommands.add.usage"))
                playerExecutor { sender, args ->
                    val warpName = args[0] as String
                    databaseService.getWarpCount(sender) { count ->
                        if (count >= configService.getInt("config.values.maxWarps")) {
                            messageService.builder(configService.getString("language.errors.maxWarpsReached")).withPrefix().send(sender)
                            return@getWarpCount
                        }
                        databaseService.warpExists(sender, warpName) { exists ->
                            if (exists) {
                                messageService.builder(configService.getString("commands.warp.subcommands.add.messages.warpAlreadyExists"))
                                    .withPrefix().tag("warp", warpName).send(sender)
                                return@warpExists
                            }
                            databaseService.createWarp(sender, warpName, sender.location)
                            messageService.builder(configService.getString("commands.warp.subcommands.add.messages.warpAdded"))
                                .withPrefix().tag("warp", warpName).send(sender)
                        }
                    }
                }
            })
            withSubcommand(CommandAPICommand(configService.getString("commands.warp.subcommands.remove.name")).apply {
                withPermission(configService.getString("commands.warp.subcommands.remove.permission"))
                withArguments(StringArgument(configService.getString("arguments.warpName")))
                withUsage(configService.getString("commands.warp.subcommands.remove.usage"))
                playerExecutor { sender, args ->
                    val warpName = args[0] as String
                    databaseService.warpExists(sender, warpName) { exists ->
                        if (exists) {
                            databaseService.deleteWarp(sender, warpName)
                            messageService.builder(configService.getString("commands.warp.subcommands.remove.messages.warpRemoved"))
                                .withPrefix().tag("warp", warpName).send(sender)
                        } else {
                            messageService.builder(configService.getString("commands.warp.subcommands.remove.messages.warpNotFound"))
                                .withPrefix().tag("warp", warpName).send(sender)
                        }
                    }
                }
            })
        }
    }
}